src/auth/apiheader/gui/qgsauthapiheaderedit.cpp

```diff
@@ -0,0 +1,188 @@
+
+  connect( btnAddHeaderPair, &QToolButton::clicked, this, &QgsAuthApiHeaderEdit::addHeaderPair );
+  connect( btnRemoveHeaderPair, &QToolButton::clicked, this, &QgsAuthApiHeaderEdit::removeHeaderPair );
+}
+
+
+bool QgsAuthApiHeaderEdit::validateConfig()
+{
+  const bool curvalid = !emptyHeadersKeysPresent() && tblwdgHeaderPairs->rowCount() != 0;
+  if ( mValid != curvalid )
+  {
+    mValid = curvalid;
+    emit validityChanged( curvalid );
+  }
+  return curvalid;
+}
+
+
+QgsStringMap QgsAuthApiHeaderEdit::configMap() const
+{
+  return headerPairs();
+}
+
+
+void QgsAuthApiHeaderEdit::loadConfig( const QgsStringMap &configmap )
+{
+  clearConfig();
+
+  mConfigMap = configmap;
+
+  populateHeaderPairs( configmap );
+
+  validateConfig();
+}
+
+
+void QgsAuthApiHeaderEdit::resetConfig()
+{
+  loadConfig( mConfigMap );
+}
+
+
+void QgsAuthApiHeaderEdit::clearConfig()
+{
+  clearHeaderPairs();
+}
+
+
+void QgsAuthApiHeaderEdit::addHeaderPair()
+{
+  addHeaderPairRow( QString(), QString() );
+  tblwdgHeaderPairs->setFocus();
+  tblwdgHeaderPairs->setCurrentCell( tblwdgHeaderPairs->rowCount() - 1, 0 );
+  tblwdgHeaderPairs->edit( tblwdgHeaderPairs->currentIndex() );
+}
+
+
+void QgsAuthApiHeaderEdit::removeHeaderPair()
+{
+  tblwdgHeaderPairs->removeRow( tblwdgHeaderPairs->currentRow() );
+  validateConfig();
+}
+
+
+void QgsAuthApiHeaderEdit::clearHeaderPairs()
+{
+  for ( int i = tblwdgHeaderPairs->rowCount(); i > 0 ; --i )
+  {
+    tblwdgHeaderPairs->removeRow( i - 1 );
+  }
+  validateConfig();
+}
+
+
+void QgsAuthApiHeaderEdit::populateHeaderPairs( const QgsStringMap &headerpairs, bool append )
+{
+  if ( !append )
+  {
+    clearHeaderPairs();
+  }
+
+  QgsStringMap::const_iterator i = headerpairs.constBegin();
+  while ( i != headerpairs.constEnd() )
+  {
+    addHeaderPairRow( i.key(), i.value() );
+    ++i;
+  }
+}
+
+
+void QgsAuthApiHeaderEdit::headerTableSelectionChanged()
+{
+  const bool hasPair = tblwdgHeaderPairs->selectedItems().count() > 0;
+  btnRemoveHeaderPair->setEnabled( hasPair );
+  validateConfig();
+}
+
+
+void QgsAuthApiHeaderEdit::headerTableCellChanged( const int row, const int column )
+{
+  Q_UNUSED( row );
+  Q_UNUSED( column );
+
+  validateConfig();
+}
+
+
+bool QgsAuthApiHeaderEdit::emptyHeadersKeysPresent()
+{
+  const int rowCount = tblwdgHeaderPairs->rowCount();
+  QgsDebugMsg( QStringLiteral( "Validate header table contains valid header keys for %1 rows" ).arg( rowCount ) );
+
+  for ( int i = 0; i < rowCount; ++i )
+  {
+    if ( tblwdgHeaderPairs->item( i, 0 )->text().isEmpty() )
+    {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+
+void QgsAuthApiHeaderEdit::addHeaderPairRow( const QString &key, const QString &val )
+{
+  const int rowCnt = tblwdgHeaderPairs->rowCount();
+  tblwdgHeaderPairs->insertRow( rowCnt );
+
+  const Qt::ItemFlags itmFlags = Qt::ItemIsEnabled | Qt::ItemIsSelectable
```

Comment:
```suggestion
  const Qt::ItemFlags itemFlags = Qt::ItemIsEnabled | Qt::ItemIsSelectable
```

abbrevations are discourage in the source -- it makes things trickier for non-English native speakers.